crypto/engine/eng_dyn.c

```diff
@@ -443,8 +443,17 @@ static int dynamic_load(ENGINE *e, dynamic_data_ctx *ctx)
          * We fail if the version checker veto'd the load *or* if it is
          * deferring to us (by returning its version) and we think it is too
          * old.
+         * Unfortunately the version checker does not distinguish between
+         * engines built for openssl 1.1.x and openssl 3.x, but loading
+         * an engine that is built for openssl 3.x will cause a fatal
+         * error.  Detect such engines, since EVP_PKEY_get_base_id is exported
+         * as a function in openssl 3.x, while it is named EVP_PKEY_base_id
+         * in openssl 1.1.x.  Therefore we take the presence of that symbol
+         * as an indication that the engine will be incompatible.
          */
-        if (vcheck_res < OSSL_DYNAMIC_OLDEST) {
+        if (vcheck_res < OSSL_DYNAMIC_OLDEST
+                || DSO_bind_func(ctx->dynamic_dso,
+                                 "EVP_PKEY_get_base_id") != NULL) {
```

Comment:
Why should 1.1.1 fail on engines built for 1.1.1?

Reply:
Isn't this checking for a 3.0 symbol?

Reply:
Yes, when this symbol is found, we can be sure it's an engine built for openssl 3.x

Reply:
... but, you're making this change *against the 1.1.1 branch*!  Why would 1.1.1 libcrypto require engines built against 3.0?

Reply:
It shall reject such engines.

Reply:
In other words, should the comparison be `== NULL` rather than `!= NULL`?

Reply:
Because we forgot to bump the API number and 1.1.1 thinks it can load 3.0 engines.

Reply:
Or wait, the comparison does seem correct here, but is then wrong in #17112?

Reply:
no I check for something different in #17112 mind the "_get_" in EVP_PKEY_get_base_id

Reply:
Ahhhhh!